### PR TITLE
Basic fcntl-style locking.

### PR DESCRIPTION
**Pull request by sunfishcode**
Add an `fcntl_lock` function implementing fcntl-style process-associated locking. Currently this only supports locking the entire file.

### CODE REVIEW

File: src/backend/linux_raw/fs/syscalls.rs [hunk 1]
```diff
@@ -1022,6 +1022,52 @@ pub(crate) fn fcntl_add_seals(fd: BorrowedFd<'_>, seals: SealFlags) -> io::Resul
...
+            l_type: l_type as _,
+            l_whence: 0,
+            l_start: 0,
+            l_len: 0,
+            ..core::mem::zeroed()
+        };
+        #[cfg(target_pointer_width = "32")]
+        {
+            ret(syscall_readonly!(
+                __NR_fcntl64,
```
Comment (cgwalters) [inline src/backend/linux_raw/fs/syscalls.rs:1053]: How about `if cfg!(target_pointer_width = "32") { __NR_fcntl64 } else { __NR_fcntl }` here and dedup the blocks?

File: src/backend/linux_raw/fs/syscalls.rs [hunk 2]
```diff
@@ -1022,6 +1022,52 @@ pub(crate) fn fcntl_add_seals(fd: BorrowedFd<'_>, seals: SealFlags) -> io::Resul
...
+        FlockOperation::NonBlockingUnlock => (F_SETLK, F_UNLCK),
+    };
+
+    unsafe {
+        let lock = flock {
+            l_type: l_type as _,
+            l_whence: 0,
+            l_start: 0,
+            l_len: 0,
+            ..core::mem::zeroed()
```
Comment (cgwalters) [inline src/backend/linux_raw/fs/syscalls.rs:1048]: Feels like we could rely on `core::mem::zeroed()` also handling the zero-init for the other members?

**Review by cgwalters — APPROVED** [review]
Somewhat superficial LGTM

File: src/backend/linux_raw/fs/syscalls.rs (hunk 1 above)
Comment (sunfishcode) [reply src/backend/linux_raw/fs/syscalls.rs:1059]: Unfortunately `syscall_readonly!` is a macro that does special things with the first parameter and needs it to be a `__NR_*` literal.

File: src/backend/linux_raw/fs/syscalls.rs (hunk 2 above)
Comment (sunfishcode) [reply src/backend/linux_raw/fs/syscalls.rs:1053]: Setting `l_len` to zero is a special case in the documented behavior of `fcntl`, so I like keeping it explicit. I've now added more comments about this though.

**Pull request merged**